{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d67392bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import io\n",
    "import pandas as pd\n",
    "import bioframe as bf\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faee27cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Settings\n",
    "in_filepath = 'data/GRCh38.UCSC.SimpleRepeats.tsv.gz'\n",
    "out_filepath = 'GRCh38.UCSC.SimpleRepeats.merged.trgt.bed'\n",
    "\n",
    "max_cluster_len = 5000 # Exclude clusters of loci > than this many bp\n",
    "join_dist = 250 # If loci are within this many bp, join them into a compound locus (no interruption kept currently)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3b715f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_gzip(file_path):\n",
    "    with open(file_path, 'rb') as file:\n",
    "        return file.read(2) == b'\\x1f\\x8b'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e223e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_intervals(bed):\n",
    "    \"\"\"Assumes sorted intervals - need to test this?\n",
    "    \n",
    "    bed is a pandas datafram with at least the columns chrom, start and end\n",
    "    \"\"\"\n",
    "    assert len(set(bed['chrom']))==1 # check all chromosomes are the same\n",
    "    \n",
    "    intervals = list(zip(bed['start'], bed['end']))\n",
    "    layers = [[intervals[0]]]\n",
    "    minstart, maxend = intervals[0]\n",
    "    \n",
    "    for interval in intervals[1:]:\n",
    "        \n",
    "        if interval[0] < minstart: minstart = interval[0]\n",
    "        if interval[1] > maxend: maxend = interval[1]\n",
    "        for layer in layers:\n",
    "            if layer[-1][-1] < interval[0]:\n",
    "                layer.append(interval)\n",
    "                break\n",
    "        else:\n",
    "            layers.append([interval])\n",
    "\n",
    "    plt.figure()\n",
    "    for i, lay in enumerate(layers):\n",
    "        x1, x2 = zip(*lay)\n",
    "        plt.hlines([i + 1] * len(x1), x1, x2, lw=30, color='lightblue')\n",
    "# To add text. This isn't working yet...\n",
    "#         print(x1, x2)\n",
    "#         for bar in zip(x1, x2):\n",
    "#             print(bar)\n",
    "#             if text is not None:\n",
    "#                 plt.text(np.mean(bar), i + 1, 'test', fontsize=12,\n",
    "#                          horizontalalignment='center', verticalalignment='center', )\n",
    "\n",
    "    plt.xlim(minstart - 2, maxend + 1)\n",
    "    plt.ylim(0, len(layers) + 1)\n",
    "\n",
    "df1 = pd.DataFrame([\n",
    "    ['chr1', 3, 5],\n",
    "    ['chr1', 4, 10],\n",
    "    ['chr1', 12, 15],\n",
    "    ],\n",
    "    columns=['chrom', 'start', 'end']\n",
    ")\n",
    "\n",
    "plot_intervals(df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f646a140",
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_gzip(in_filepath):\n",
    "    bed_file = io.TextIOWrapper(gzip.open(in_filepath, 'rb'), encoding='utf-8')\n",
    "else:\n",
    "    bed_file = open(in_filepath)\n",
    "        \n",
    "bed = pd.read_table(bed_file)\n",
    "bed.drop(columns='#bin', inplace=True)\n",
    "bed.rename(columns = {'chromStart':'start', 'chromEnd': 'end'}, inplace=True)\n",
    "# print(bed)\n",
    "assert bf.core.checks.is_bedframe(bed)\n",
    "\n",
    "print('Rows in bed: ', len(bed))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60e3f0d8",
   "metadata": {},
   "source": [
    "Parse and cluster bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0c6a237",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "bedclusters = bf.cluster(bed, min_dist=None)\n",
    "print('Number of clusters:', len(bedclusters['cluster'].unique()))\n",
    "bedclusters.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdbfce9e",
   "metadata": {},
   "source": [
    "## Plot some metrics about genome annotation\n",
    "\n",
    "### Gaps between loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b7d3fed",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed['direction'] = '.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21534b34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Calculate distance to closest repeat\n",
    "# closest_intervals = bf.closest(bed, df2=None, return_distance=True, direction_col = 'direction')\n",
    "# display(closest_intervals)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32e277ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf42886",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_overlaps(clusterbed, pct_overlap = 5):\n",
    "    \n",
    "    if len(clusterbed) == 1:\n",
    "        return(clusterbed)\n",
    "    \n",
    "    # If multiple, take the largest\n",
    "    maxlen = (clusterbed['end'] - clusterbed['start']).max()\n",
    "    best = clusterbed[clusterbed['end'] - clusterbed['start'] == maxlen]\n",
    "    \n",
    "    # If tied, take the highest percent match\n",
    "    if len(best) > 1:\n",
    "        max_match = (best['perMatch']).max()\n",
    "        best = best[best['perMatch'] == max_match]\n",
    "\n",
    "    # If still tied, take the smallest period\n",
    "    if len(best) > 1:\n",
    "        minperiod = (best['period']).min()\n",
    "        best = best[best['period'] == minperiod]\n",
    "\n",
    "    # If stll tied, take the highest score\n",
    "    if len(best) > 1:\n",
    "        maxscore = (best['score']).max()\n",
    "        best = best[best['score'] == maxscore]\n",
    "        \n",
    "    # If still tied, take the first\n",
    "    if len(best) > 1:\n",
    "        best = best.head(1)\n",
    "    \n",
    "    assert len(best) == 1\n",
    "    \n",
    "    # Check if any intervals remain that don't overlap the largest, or only overlap by less than x%\n",
    "    # Not sure if this is needed at all?\n",
    "#     overlaps = bf.coverage(clusterbed, best)\n",
    "#     overlaps['pct_overlap'] = overlaps['coverage']/(clusterbed['end'] - clusterbed['start'])*100\n",
    "#     alsokeep = overlaps[overlaps['pct_overlap'] < pct_overlap]\n",
    "    \n",
    "    # Trim overlapping sequence\n",
    "    trimmed = bf.subtract(clusterbed, best)\n",
    "\n",
    "    if len(trimmed) > 0:\n",
    "        return bf.sort_bedframe(pd.concat([best, trimmed], join='inner', ignore_index=True, copy=True))\n",
    "    else:\n",
    "        return best.copy()\n",
    "\n",
    "\n",
    "# Tests\n",
    "\n",
    "# HD locus example\n",
    "# df1 = pd.DataFrame([\n",
    "#     ['chr4', 3074877, 3074940, 3, 119],\n",
    "#     ['chr4', 3074877, 3074940, 3, 200],\n",
    "#     ['chr4', 3074877, 3074940, 2, 200],\n",
    "#     ['chr4', 3074939, 3074968, 3, 51],\n",
    "#     ['chr4', 3074970, 3074977, 3, 51],\n",
    "#     ],\n",
    "#     columns=['chrom', 'start', 'end', 'period', 'score']\n",
    "# )\n",
    "\n",
    "# Failed first attempt (good test case)\n",
    "# df1 = pd.DataFrame([\n",
    "#     ['chr1', 12582784, 12583841, 3, 119],\n",
    "#     ['chr1', 12583317, 12583892, 3, 119],\n",
    "#     ],\n",
    "#     columns=['chrom', 'start', 'end', 'period', 'score']\n",
    "# )\n",
    "\n",
    "# One interval inside other\n",
    "# df1 = pd.DataFrame([\n",
    "#     ['chr1', 154140082, 154141019, 3, 119],\n",
    "#     ['chr1', 154140668, 154140702, 3, 119],\n",
    "#     ],\n",
    "#     columns=['chrom', 'start', 'end', 'period', 'score']\n",
    "# )\n",
    "# plot_intervals(df1)\n",
    "# print(fix_overlaps(df1))\n",
    "# plot_intervals(fix_overlaps(df1))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "039a911a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "n_big = 0\n",
    "\n",
    "# Fix overlapping intervals by choosing the largest then trimming the others\n",
    "all_list = []\n",
    "for cluster in bedclusters['cluster'].unique():\n",
    "    clusterbed = bedclusters.loc[bedclusters['cluster'] == cluster]\n",
    "    \n",
    "    if clusterbed.iloc[0]['cluster_end'] - clusterbed.iloc[0]['cluster_start'] > max_cluster_len:\n",
    "        n_big += 1\n",
    "        continue\n",
    "\n",
    "    if len(clusterbed) == 1:\n",
    "        all_list.append(clusterbed)\n",
    "        \n",
    "    if len(clusterbed) > 1:\n",
    "        non_overlapping = fix_overlaps(clusterbed)\n",
    "        all_list.append(non_overlapping)\n",
    "    \n",
    "all_non_overlappping = pd.concat(all_list, ignore_index=True, copy=True)\n",
    "    \n",
    "print('skipped due to size: ', n_big)\n",
    "all_non_overlappping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05339268",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster loci within x bp of each other\n",
    "\n",
    "# Drop old cluster columns\n",
    "all_non_overlappping.drop(columns=['cluster', 'cluster_start', 'cluster_end'], inplace = True)\n",
    "bedclusters2 = bf.cluster(bed, min_dist=join_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f869f48d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_loci(clusterbed):\n",
    "    \"\"\"Join simple repeats into compound loci\n",
    "    Non-repetitive sequence between them will be ignored.\n",
    "    \n",
    "    Output format example string: \n",
    "    chr1\t57367043\t57367119\tID=chr1_57367043_57367119;MOTIFS=AAAAT,GAAAT;STRUC=(AAAAT)n(GAAAT)n(AAAAT)n\n",
    "    \"\"\"\n",
    "\n",
    "    assert len(set(clusterbed['chrom']))==1 # check all chromosomes are the same\n",
    "    # Check that none overlap?\n",
    "    # Check sorted?\n",
    "    \n",
    "    motifs = clusterbed['sequence']\n",
    "    \n",
    "    uniq_motifs = ','.join(list(dict.fromkeys(motifs))) # Requires Python 3.7+ to maintain ordering\n",
    "    struc = ''.join([f'({x})n'.format() for x in motifs])\n",
    "\n",
    "    chrom = clusterbed['chrom'].values[0]\n",
    "    start = clusterbed['cluster_start'].values[0]\n",
    "    end = clusterbed['cluster_end'].values[0]\n",
    "\n",
    "    loc_id = f'{chrom}_{start}_{end}'.format()\n",
    "    outline = f'{chrom}\\t{start}\\t{end}\\tID={loc_id};MOTIFS={uniq_motifs};STRUC={struc}'.format()\n",
    "    return outline\n",
    "    \n",
    "# Tests\n",
    "\n",
    "# HD locus example (ish)\n",
    "df1 = pd.DataFrame([\n",
    "    ['chr4', 3074877, 3074940, 'CAG', 3074877, 3074990],\n",
    "    ['chr4', 3074939, 3074968, 'GCC', 3074877, 3074990],\n",
    "    ['chr4', 3074975, 3074990, 'AT', 3074877, 3074990],\n",
    "    ],\n",
    "    columns=['chrom', 'start', 'end', 'sequence', 'cluster_start', 'cluster_end']\n",
    ")\n",
    "\n",
    "plot_intervals(df1)\n",
    "print(merge_loci(df1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9152c7cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join nearby loci into compound ones\n",
    "n_big = 0\n",
    "merged = 0\n",
    "max_merge = 0\n",
    "biggest_cluster = None\n",
    "\n",
    "metrics = open('metrics.txt', 'w')\n",
    "metrics.write('chrom\\tstart\\tlength\\tmotifs\\n')\n",
    "with open(out_filepath, 'w') as outfile:\n",
    "    for cluster in bedclusters2['cluster'].unique():\n",
    "        clusterbed = bedclusters2.loc[bedclusters2['cluster'] == cluster]\n",
    "\n",
    "        chrom = clusterbed['chrom'].values[0]\n",
    "        start = clusterbed['cluster_start'].values[0]\n",
    "        length = clusterbed['cluster_end'].values[0] - clusterbed['cluster_start'].values[0]\n",
    "        motifs = len(clusterbed)\n",
    "        metrics.write(f'{chrom}\\t{start}\\t{length}\\t{motifs}\\n')\n",
    "        \n",
    "        if clusterbed['cluster_end'].values[0] - clusterbed['cluster_start'].values[0] > max_cluster_len:\n",
    "            n_big += 1\n",
    "            continue\n",
    "        \n",
    "        if len(clusterbed) > 1:\n",
    "            merged += 1\n",
    "            if len(clusterbed) > max_merge:\n",
    "                max_merge = len(clusterbed)\n",
    "                biggest_cluster = clusterbed\n",
    "\n",
    "        outfile.write(merge_loci(clusterbed) + '\\n')\n",
    "        \n",
    "        \n",
    "            #all_merged = pd.concat([all_merged, merged_loci], ignore_index=True, copy=True)\n",
    "    \n",
    "print('skipped due to size: ', n_big)\n",
    "print('This many loci are compound: ', merged)\n",
    "biggest_cluster\n",
    "#all_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "076806fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If any cluster overlap a known pathogenic locus, replace them with the manually curated pathogenic locus"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
